### ISSUE TITLE
Using HAST for describing a markdown document

### ISSUE THREAD

**Issue by davidbonnet** [syntax-tree/unist#12]
I like a lot the Unist ecosystem and its AST-oriented approach.

However, I do not understand the motivation behind creating two widely different ASTs for markdown and HTML documents. I would expect HAST to include MDAST (as shown below) and therefore an HTML parser to produce an AST that a markdown compiler would understand, without requiring a transformation step (e.g., with `rehype-remark`). Conversely, a markdown parser should be able to produce an AST that an HTML compiler would understand.

In other words, would it make sense to build a `rehype-stringify-markdown` and a `rehype-parse-markdown` and ignore the MDAST? Is there something that would prevent that?

---

For example, this MDAST node:
```js
{
  type: "paragraph",
  children: [{
    type: "text",
    value: "Hello!"
  }]
}
```
…contains at most as much information as this HAST node:
```js
{
  type: "element",
  tagName: "p",
  properties: {},
  children: [{
    type: "text",
    value: "Hello!"
  }]
}
```

**Comment by wooorm** [comment]
Hi @davidbonnet! Thanks for the detailed issue.

They are similar, yes, that‘s why the conversion isn’t very hard. But they’re not the same, and need special handling, that’s why the conversion is needed. HAST stores more and less information than MDAST, and the same goes the other way around.

For example, markdown supports extended stuff in the code flag: `~~~js id=1\ncode()\n~~~`, which is not exposed to HTML anywhere. Markdown also supports link definitions, that may never show up in HTML.

The whole point of remark is that markdown isn’t just used for going to HTML. It also goes to latex, man pages, terminals, other stuff, and back to markdown.

Also, the whole ecosystem is pluggable. You could create `rehype-parse-markdown` yourself, with something like this:

```js
module.exports = rehypeParseMarkdown;
var p = new remarkParse.Parser();
function rehypeParseMarkdown() {
  this.Parser = parser;
  function parser(file) {
    return mdast2hast(p.parse(file));
  }
}
```

...but I advise against it, it’s not *that* hard to `.use(remarkParse).use(remark2rehype)`.

Finally, you could use any markdown to HTML compiler out there, and use rehype later on its output!
